File: crates/harness-router/src/tier.rs
//! Difficulty tiers + the tier→model routing table.

use std::collections::HashMap;

/// Difficulty tier a single turn is classified into. Mirrors the
/// four-level scheme used by comparable routers: trivial confirmations
/// at `Simple`, deep multi-file analysis at `Reasoning`.
///
/// The [`HeuristicClassifier`](crate::HeuristicClassifier) only emits
/// `Simple` / `Medium` / `Complex`; `Reasoning` is reserved for the
/// future LLM judge but is a valid routing target today (operators may
/// map it even if nothing emits it yet).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Tier {
    Simple,
    Medium,
    Complex,
    Reasoning,
}

impl Tier {
    /// Stable lowercase wire/log name.
    pub fn as_str(&self) -> &'static str {
        match self {
            Tier::Simple => "simple",
            Tier::Medium => "medium",
            Tier::Complex => "complex",
            Tier::Reasoning => "reasoning",
        }
    }
}

/// One `(provider, model)` routing target. `provider` is a key into the
/// [`RoutingProvider`](crate::RoutingProvider)'s provider map; `model` is
/// the string written onto the outgoing request.
///
/// The `provider/model` slash form matches the rest of the harness
/// (`ProviderRegistry`, `[routing]` config), so operators reuse the same
/// syntax in `JARVIS_ROUTER_TIER_*` env vars.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ModelRef {
    pub provider: String,
    pub model: String,
}

impl ModelRef {
    pub fn new(provider: impl Into<String>, model: impl Into<String>) -> Self {
        Self {
            provider: provider.into(),
            model: model.into(),
        }
    }

    /// Parse the `<provider>/<model>` wire form. Returns `None` when
    /// either side is empty so a typo'd env var falls through to the
    /// default target rather than routing to a half-specified model.
    pub fn parse(s: &str) -> Option<Self> {
        let (p, m) = s.split_once('/')?;
        if p.is_empty() || m.is_empty() {
            return None;
        }
        Some(Self::new(p, m))
    }
}

/// Tier→model table with a mandatory `default` fallback. Any tier with
/// no explicit entry resolves to `default`, so a partially-configured
/// router (e.g. only `Simple` mapped) is always well-defined.
#[derive(Debug, Clone)]
pub struct RouterConfig {
    pub default: ModelRef,
    pub tiers: HashMap<Tier, ModelRef>,
}

impl RouterConfig {
    pub fn new(default: ModelRef) -> Self {
        Self {
            default,
            tiers: HashMap::new(),
        }
    }

    /// Builder: map `tier` to `target`. Later calls overwrite earlier
    /// ones for the same tier.
    pub fn with_tier(mut self, tier: Tier, target: ModelRef) -> Self {
        self.tiers.insert(tier, target);
        self
    }

    /// Resolve a tier to its target, falling back to `default`.
    pub fn target(&self, tier: Tier) -> &ModelRef {
        self.tiers.get(&tier).unwrap_or(&self.default)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_accepts_slash_form() {
        let r = ModelRef::parse("anthropic/claude-opus").unwrap();
        assert_eq!(r.provider, "anthropic");
        assert_eq!(r.model, "claude-opus");
    }

    #[test]
    fn parse_rejects_empty_segments() {
        assert!(ModelRef::parse("openai/").is_none());
        assert!(ModelRef::parse("/m").is_none());
        assert!(ModelRef::parse("noslash").is_none());
    }

    #[test]
    fn target_falls_back_to_default() {
        let cfg = RouterConfig::new(ModelRef::new("openai", "gpt-4o-mini"))
            .with_tier(Tier::Complex, ModelRef::new("anthropic", "claude-opus"));
        assert_eq!(cfg.target(Tier::Complex).model, "claude-opus");
        // Medium is unset → default.
        assert_eq!(cfg.target(Tier::Medium).model, "gpt-4o-mini");
    }
}

File: packages/router/src/tier.ts
// Difficulty tiers + the tier→model routing table.
// Ported from harness-router/src/tier.rs.
//
// The Rust `Tier` enum is rendered as a lowercase string union here — its
// stable wire/log names (`as_str`) ARE the variants, so no separate mapping
// helper is needed. `RouterConfig` keeps the mandatory `default` fallback so a
// partially-configured router is always well-defined.

/**
 * Difficulty tier a single turn is classified into. Mirrors the four-level
 * scheme used by comparable routers: trivial confirmations at "simple", deep
 * multi-file analysis at "reasoning".
 *
 * The HeuristicClassifier only emits "simple" / "medium" / "complex";
 * "reasoning" is reserved for the future LLM judge but is a valid routing
 * target today (operators may map it even if nothing emits it yet).
 */
export type Tier = "simple" | "medium" | "complex" | "reasoning";

/** All tiers, in difficulty order. Stable lowercase wire/log names. */
export const ALL_TIERS: readonly Tier[] = ["simple", "medium", "complex", "reasoning"];

/**
 * One `(provider, model)` routing target. `provider` is a key into the
 * RoutingProvider's provider map; `model` is the string written onto the
 * outgoing request.
 *
 * The `provider/model` slash form matches the rest of the harness, so
 * operators reuse the same syntax in `JARVIS_ROUTER_TIER_*` env vars.
 */
export interface ModelRef {
  provider: string;
  model: string;
}

export function modelRef(provider: string, model: string): ModelRef {
  return { provider, model };
}

/**
 * Parse the `<provider>/<model>` wire form. Returns `undefined` when either
 * side is empty so a typo'd env var falls through to the default target rather
 * than routing to a half-specified model. Splits on the FIRST `/` so model
 * names containing slashes are preserved (mirrors Rust `split_once`).
 */
export function parseModelRef(s: string): ModelRef | undefined {
  const idx = s.indexOf("/");
  if (idx === -1) return undefined;
  const provider = s.slice(0, idx);
  const model = s.slice(idx + 1);
  if (provider === "" || model === "") return undefined;
  return { provider, model };
}

/**
 * Tier→model table with a mandatory `default` fallback. Any tier with no
 * explicit entry resolves to `default`, so a partially-configured router (e.g.
 * only "simple" mapped) is always well-defined.
 *
 * Ported from the Rust `RouterConfig` struct + builder. `withTier` returns a
 * new config (immutable builder) so the same base can be reused.
 */
export class RouterConfig {
  readonly default: ModelRef;
  readonly tiers: Map<Tier, ModelRef>;

  constructor(defaultTarget: ModelRef, tiers?: Map<Tier, ModelRef>) {
    this.default = defaultTarget;
    this.tiers = tiers ?? new Map();
  }

  /**
   * Map `tier` to `target`. Later calls overwrite earlier ones for the same
   * tier. Returns a new RouterConfig (builder style) leaving the receiver
   * untouched.
   */
  withTier(tier: Tier, target: ModelRef): RouterConfig {
    const next = new Map(this.tiers);
    next.set(tier, target);
    return new RouterConfig(this.default, next);
  }

  /** Resolve a tier to its target, falling back to `default`. */
  target(tier: Tier): ModelRef {
    return this.tiers.get(tier) ?? this.default;
  }
}
